tests/cli/test_list.py

```diff
@@ -1,96 +1,161 @@
 
 from pdm.cli import actions
-from pdm.exceptions import PdmException
+from pdm.cli.commands.list import Command
+from pdm.models.specifiers import PySpecSet
+from tests import FIXTURES
+from tests.conftest import Distribution
 
 
 def test_list_command(project, invoke, mocker):
-    do_list = mocker.patch.object(actions, "do_list")
+    # Calls the correct handler within the Command
+    m = mocker.patch.object(Command, "handle_list")
     invoke(["list"], obj=project)
-    do_list.assert_called_once()
+    m.assert_called_once()
 
 
 @pytest.mark.usefixtures("repository", "working_set")
-def test_list_dependency_graph(project, capsys):
+def test_list_graph_command(project, invoke, mocker):
+    # Calls the correct handler within the list command
+    m = mocker.patch.object(Command, "handle_graph")
+    invoke(["list", "--graph"], obj=project)
+    m.assert_called_once()
+
+
+@mock.patch("rich.console.ConsoleOptions.ascii_only", lambda: True)
+@pytest.mark.usefixtures("repository", "working_set")
+def test_list_dependency_graph(project, invoke):
+    # Shows a line that contains a sub requirement (any order).
     actions.do_add(project, packages=["requests"])
-    actions.do_list(project, graph=True)
-    content, _ = capsys.readouterr()
-    assert "└── urllib3 1.22 [ required: <1.24,>=1.21.1 ]" in content
+    result = invoke(["list", "--graph"], obj=project)
+    expected = "-- urllib3 1.22 [ required: <1.24,>=1.21.1 ]" in result.outputs
+    assert expected
+
+
+@mock.patch("rich.console.ConsoleOptions.ascii_only", lambda: True)
+@pytest.mark.usefixtures("repository", "working_set")
+def test_list_dependency_graph_include_exclude(project, invoke):
+    # Just include dev packages in the graph
+    project.environment.python_requires = PySpecSet(">=3.6")
+    dep_path = FIXTURES.joinpath("projects/demo").as_posix()
+    actions.do_add(
+        project,
+        dev=True,
+        group="dev",
+        editables=[f"{dep_path}[security]"],
+    )
+
+    # Output full graph
+    # TODO: Find out why the chardet and idna versions are different?
+    result = invoke(["list", "--graph"], obj=project)
+    expects = (
+        "demo 0.0.1 [ Not required ]\n",
+        "+-- chardet 3.0.4 [ required: Any ]\n",
+        "`-- idna 2.7 [ required: Any ]\n",
+        "requests 2.19.1 [ Not required ]\n",
+        "+-- certifi 2018.11.17 [ required: >=2017.4.17 ]\n",
+        "+-- chardet 3.0.4 [ required: <3.1.0,>=3.0.2 ]\n",
+        "+-- idna 2.7 [ required: <2.8,>=2.5 ]\n",
+        "`-- urllib3 1.22 [ required: <1.24,>=1.21.1 ]\n",
+    )
+    expects = "".join(expects)
+    assert expects == result.outputs
+
+    # Only include the dev dep
+    result = invoke(["list", "--graph", "--include", "dev"], obj=project)
+    expects = (
+        "demo 0.0.1 [ Not required ]\n",
+        "+-- chardet [ not installed ] [ required: Any ]\n",  # installed 3.0.4 ?
+        "`-- idna [ not installed ] [ required: Any ]\n",  # installed 2.7 ?
```

Comment:
Here the `chardet` and `idna` packages are showing as `[ not installed ]`.   I am not sure why this is showing like this, because `--resolve` is not used, so these packages should all be from the working set.

Reply:
I explicitly added `idna` and `chardet` them as dev deps in the test and made a note.  I think for the purpose of the test it is reasonable/clearer than a complicated mock.